src/dynamic_batch_scheduler.cc

```diff
@@ -598,29 +598,34 @@ DynamicBatchScheduler::DelegateResponse(
   std::lock_guard<std::mutex> lock(completion_queue_mtx_);
   completion_queue_.emplace_back();
   auto queue_slot = &completion_queue_.back();
-  uint64_t request_hash = request->CacheKey();
   // Pass raw ptr to lambda for tracking stats from cache and updating
   // metric reporter on cache miss stats after insertion
   InferenceRequest* raw_request_ptr = request.get();
 
   request->SetResponseDelegator(
-      [this, queue_slot, request_hash, raw_request_ptr](
+      [this, queue_slot, raw_request_ptr](
           std::unique_ptr<InferenceResponse>&& response, const uint32_t flags) {
-        if (response_cache_enabled_) {
+        if (response_cache_enabled_ && raw_request_ptr->CacheKeyIsSet()) {
           // Cache insertion happens here because we need the backend to have
           // computed the inference response first in the case of cache miss
           auto cache = model_->Server()->GetResponseCache();
-          auto status = cache->Insert(request_hash, *response, raw_request_ptr);
-          if (status.IsOk()) {
+          auto status = cache->Insert(*response, raw_request_ptr);
+          // Cache Hit
+          if (status.StatusCode() == Status::Code::ALREADY_EXISTS) {
+            // Do nothing as we already update cache hit statistics on Lookup
+          }
+          // Cache Miss or failure to insert
+          else {
 #ifdef TRITON_ENABLE_STATS
-            // Cache miss
+            // Update cache miss statistics even on failure to insert
+            // as we still spend time on lookup and attempting to insert
             raw_request_ptr->ReportStatisticsCacheMiss(reporter_.get());
 #endif  // TRITON_ENABLE_STATS
-          } else if (status.StatusCode() == Status::Code::ALREADY_EXISTS) {
-            /* Cache hit */
-          } else {
-            LOG_ERROR << "Failed to insert request_hash [" << request_hash
-                      << "] into response cache: " << status.Message();
+            if (!status.IsOk()) {
+              LOG_ERROR << "Failed to insert request_hash ["
+                        << raw_request_ptr->CacheKey()
+                        << "] into response cache: " << status.Message();
+            }
```

Comment:
Re-organized this to update cache miss statistics even on failure to insert, because we still spend time on lookup and insertion even if insertion fails.

Here's an example of repeatedly trying to insert a request that is larger than the total cache size:

Before:
```
  Server: 
    Inference count: 1080
    Execution count: 1080
    Cache hit count: 0
    Cache miss count: 0
    Successful request count: 1080
    Avg request latency: 4695 usec (overhead 215 usec + queue 2537 usec + cache hit/miss 1943 usec)
      Average Cache Hit Latency: 0 usec
      Average Cache Miss Latency: 1943 usec (cache lookup/insertion 0 usec + compute input 343 usec + compute infer 1544 usec + compute output 56 usec)
```

After always updating cache miss even if insert fails:
```
  Server: 
    Inference count: 1132
    Execution count: 1132
    Cache hit count: 0
    Cache miss count: 1132
    Successful request count: 1132
    Avg request latency: 4460 usec (overhead 0 usec + queue 2377 usec + cache hit/miss 4195 usec)
      Average Cache Hit Latency: 0 usec
      Average Cache Miss Latency: 4193 usec (cache lookup/insertion 2324 usec + compute input 337 usec + compute infer 1482 usec + compute output 50 usec)
```

I noticed total average latency was not equal to the sum of it's parts here, so I noticed the total request_duration isn't being updated to include both lookup+insertion on misses. After fixing that:
```
  Server: 
    Inference count: 1135
    Execution count: 1135
    Cache hit count: 0
    Cache miss count: 1135
    Successful request count: 1135
    Avg request latency: 6733 usec (overhead 215 usec + queue 2353 usec + cache hit/miss 4165 usec)
      Average Cache Hit Latency: 0 usec
      Average Cache Miss Latency: 4165 usec (cache lookup/insertion 2302 usec + compute input 346 usec + compute infer 1467 usec + compute output 49 usec)
```

CC @tanmayv25 @matthewkotila for a sanity check.